src/emulator/hardware/builtin-gates/MipsAlu.js

```diff
@@ -0,0 +1,92 @@
+/**
+ * The MIT License (MIT)
+ * Copyright (c) 2017-present Dmitry Soshnikov <dmitry.soshnikov@gmail.com>
+ */
+
+'use strict';
+
+const BuiltInGate = require('../BuiltInGate');
+
+/**
+ * Canonical truth table for the MipsAlu
+ */
+const TRUTH_TABLE = [
+
+  // Add
+  {a: 1, b: 1, na: 0, nb: 0, less: 0, cin: 0, op: 0b10, out: 0, cout: 1, set: 0 },
+  {a: 1, b: 1, na: 0, nb: 0, less: 0, cin: 1, op: 0b10, out: 1, cout: 1, set: 0 },
+  {a: 0, b: 0, na: 0, nb: 0, less: 0, cin: 1, op: 0b10, out: 1, cout: 0, set: 0 },
+
+  // Sub
+  {a: 1, b: 1, na: 0, nb: 1, less: 0, cin: 1, op: 0b10, out: 0, cout: 1, set: 0 },
+
+  // And
+  {a: 1, b: 1, na: 0, nb: 0, less: 0, cin: 0, op: 0b00, out: 1, cout: 0, set: 0 },
+
+  // Or
+  {a: 1, b: 1, na: 0, nb: 0, less: 0, cin: 0, op: 0b01, out: 1, cout: 0, set: 0 },
+  {a: 0, b: 1, na: 0, nb: 0, less: 0, cin: 0, op: 0b01, out: 1, cout: 0, set: 0 },
+  {a: 0, b: 0, na: 0, nb: 0, less: 0, cin: 0, op: 0b01, out: 0, cout: 0, set: 0 },
+
+];
+
+/**
+ * A 1-bit MipsAlu
+ */
+class MipsAlu extends BuiltInGate {
+  /**
+   * Logic:
+   *
+   *   a & b: op = 0
+   *   a | b: op = 1
+   *   a NOR b => ~a & ~b: op = 0, na = 1, nb = 1
+   *
+   * Math:
+   *
+   *   a + b: op = 2
+   *   a - b: op = 2, nb = 1, cin = 1
+   *   b - a: op = 2, na = 1, cin = 1
+   *
+   */
+  eval() {
+    const [ a, b, na, nb, less, cin, op ] = this.getInputPins().map(p => p.getValue());
+
+    const A = na ? +!a : a;
+    const B = nb ? +!b : b;
+
+    switch(op) {
+      case 0b00: {
+        this.getOutputPins()[0].setValue(A & B); 
+        break;
+      }
+      case 0b01: {
+        this.getOutputPins()[0].setValue(A | B); 
+        break;
+      }
+      case 0b10: {
+        const sum = A + B + cin;
+        this.getOutputPins()[0].setValue(sum % 2); 
+        this.getOutputPins()[1].setValue(Math.floor(sum / 2)); 
+        break;
+      }
+      case 0b11: {
+        // TODO: Not sure what 'less' is for yet, this is just a placeholder for eslint
+        this.getOutputPins()[0].setValue(less);
```

Comment:
This is actually correct implementation, we can remove the TODO comment.

In 1-bit ALU the `less` input is just directly propagated to the output w/o any calculations. So specifically in the 1-bit ALU it's kinda "usless".

The actual purpose of the `less` though is to be used in the 16-bit ALU. This is to support `slt` MIPS Assembly instruction. You can read more about it in the "Tailoring the 32-Bit ALU to MIPS" section of the `ALU_AppB.pdf` attached in the issue. It starts with:

> One instruction that still needs support is the set on less than instruction (slt). Recall that the operation produces 1 if rs < rt, and 0 otherwise.

I.e. if `a < b`, we should set `out` to `1`, otherwise to `0`.

`a < b`, `a = 0`, `b = 1`, `cin = 1`, `nb = 1`

```js
out = 0b0000000000000001
```

`a > b` (i.e. `!(a < b)`), `a = 1`, `b = 0`, `cin = 1`, `nb = 1`

```js
out = 0b0000000000000000
```

Technically this is done via _subtraction_, hence `cin = 1`, and `nb = 1`. From the doc:

> (a – b) < 0 ⇒ ((a – b) + b) < (0 + b) ⇒a<b

Since this is 16-bit machine, as you can see, the bits 1-15 are always `0`, and only the LSB should be either `1` or `0`.

So when 16 1-bit ALU are combined together, only the _LSB (least-significant bit)_ will be [calculated from the `set` output](https://github.com/DmitrySoshnikov/hdl-js/blob/master/examples/MipsAlu16.hdl#L26) of the `FullAdder`, see `less=set`. Bits 1-15 are [always set to `0`](https://github.com/DmitrySoshnikov/hdl-js/blob/master/examples/MipsAlu16.hdl#L27-L41) (i.e. `less=false`).

Anyways, in 1-bit ALU we just need to propagate it to the output "as is" 😃